stl/inc/iterator

```diff
@@ -350,30 +366,33 @@ public:
     using difference_type   = typename _Traits::off_type;
     using pointer           = const _Elem*;
     using reference         = _Elem;
-
-    using char_type      = _Elem;
-    using traits_type    = _Traits;
-    using streambuf_type = basic_streambuf<_Elem, _Traits>;
-    using istream_type   = basic_istream<_Elem, _Traits>;
-
-    using int_type = typename traits_type::int_type;
+    using char_type         = _Elem;
+    using traits_type       = _Traits;
+    using int_type          = typename traits_type::int_type;
+    using streambuf_type    = basic_streambuf<_Elem, _Traits>;
+    using istream_type      = basic_istream<_Elem, _Traits>;
 
     constexpr istreambuf_iterator() noexcept : _Strbuf(nullptr), _Got(true), _Val() {}
-
-    istreambuf_iterator(streambuf_type* _Sb) noexcept : _Strbuf(_Sb), _Got(!_Sb), _Val() {}
+#ifdef __cpp_lib_concepts
+    constexpr istreambuf_iterator(default_sentinel_t) noexcept : _Strbuf(nullptr), _Got(true), _Val() {}
+#endif // __cpp_lib_concepts
 
     istreambuf_iterator(istream_type& _Istr) noexcept : _Strbuf(_Istr.rdbuf()), _Got(!_Strbuf), _Val() {}
 
+    istreambuf_iterator(streambuf_type* _Sb) noexcept : _Strbuf(_Sb), _Got(!_Sb), _Val() {}
+
 private:
     class _Istreambuf_proxy {
     public:
-        _NODISCARD _Elem operator*() const {
+        _NODISCARD _Elem operator*() const noexcept(is_nothrow_copy_constructible_v<_Elem>) /* strengthened */ {

```

Comment:
I am confused by the exception specifications in `istreambuf_iterator` and its proxy. `istreambuf_iterator` stores `_Elem _Val` (not mentioned as an exposition-only data member). The Standard depicts `istreambuf_iterator` as having a `noexcept` default constructor *and* a `noexcept` copy constructor. This seems to suggest that `_Elem` must be nothrow default constructible and nothrow copy constructible. Yet here, you're querying `is_nothrow_copy_constructible_v<_Elem>`.

Should these operations be assumed to be nothrow because this is a "character type"? (I believe the answer is yes.)

Reply:
The Standard is murky here since it doesn't depict `istreambuf_iterator` as having an exposition-only `charT` member: it's possible to implement `istreambuf_iterator` by grabbing a fresh character with `sbuf_->sgetc()` on each call to `operator*` and not memoizing the result within the `istreambuf_iterator` itself. Our justification for `noexcept` here relies on the fact that `istreambuf_iterator<charT, Traits>` uses `istream<charT, Traits>`, and `istream`'s first template argument must "represent a member of the set of types containing `char`, `wchar_t`, and any other implementation-defined character types that meet the requirements for a character on which any of the iostream components can be instantiated." per one of the most underspecified bits of standardese [iostreams.limits.pos]/2.

"the requirements for a character on which any of the iostream components can be instantiated" is nearly circular, and AFAICT effectively means "types an implementation supports here which we don't require implementations to document." It's far from explicit that such types must be trivially copyable and literal as needed to support [istreambuf.iterator.general]/1 "All specializations of `istreambuf_iterator` shall have a trivial copy constructor, a `constexpr` default constructor, and a trivial destructor" if stored as a member of `istreambuf_iterator`. This is far beyond the threshold of vagueness where I expect misuse in the wild. 

Altogether the result is that I'm being cautious with `noexcept` strengthening here. I'd be happier making these unconditional if we were to `static_assert` that `charT` is trivially copyable and nothrow-default-constructible in the body of `istreambuf_iterator`, but I don't know that it's worth the potential breakage of code that's misusing this iterator in the wild.

Reply:
Cautious is fine, thanks.